src/appshell/view/appmenumodel.cpp

```diff
@@ -229,36 +226,52 @@ MenuItem AppMenuModel::editItem()
 MenuItem AppMenuModel::viewItem()
 {
     MenuItemList viewItems {
-        makeAction("toggle-palette", isNotationPage(), notationPageState()->isPanelVisible(PanelType::Palette)),
-        makeAction("toggle-instruments", isNotationPage(), notationPageState()->isPanelVisible(PanelType::Instruments)),
-        makeAction("masterpalette", isNotationPage(), paletteController()->isMasterPaletteOpened().val),
-        makeAction("inspector", isNotationPage(), notationPageState()->isPanelVisible(PanelType::Inspector)),
-        makeAction("toggle-playpanel", isNotationPage()), // need implement
-        makeAction("toggle-navigator", isNotationPage(), notationPageState()->isPanelVisible(PanelType::NotationNavigator)),
-        makeAction("toggle-timeline", isNotationPage()), // need implement
-        makeAction("toggle-mixer", isNotationPage()), // need implement
-        makeAction("synth-control", isNotationPage()), // need implement
-        makeAction("toggle-selection-window", isNotationPage()), // need implement
-        makeAction("toggle-piano", isNotationPage()), // need implement
-        makeAction("toggle-scorecmp-tool", isNotationPage()), // need implement
+        makeAction("toggle-palette", isPanelVisible(PanelType::Palette),
+                   [this]() { return m_viewMenuController->isPanelAvailable(PanelType::Palette); }),
+        makeAction("toggle-instruments", isPanelVisible(PanelType::Instruments),
+                   [this]() { return m_viewMenuController->isPanelAvailable(PanelType::Instruments); }),
+        makeAction("masterpalette", paletteController()->isMasterPaletteOpened().val,
+                   [this]() { return m_viewMenuController->isMasterPaletteAvailable(); }),
+        makeAction("inspector", isPanelVisible(PanelType::Inspector),
+                   [this]() { return m_viewMenuController->isPanelAvailable(PanelType::Inspector); }),
+        makeAction("toggle-navigator", isPanelVisible(PanelType::NotationNavigator),
+                   [this]() { return m_viewMenuController->isPanelAvailable(PanelType::NotationNavigator); }),
+        makeAction("toggle-timeline", isPanelVisible(PanelType::TimeLine),
```

Comment:
I think, we can simplify getting of availability of a menu item. You already have an abstract interface for this - IActionsAvailability. We can move this check inside makeAction:

makeAction(actionCode) {
...
item.enabled = resolveController(actionCode)->isActionAvailable(actionCode)
....
}

Reply:
We can not resolve controller in makeAction by action code because makeAction do not know about parent of action(File, View, etc)